stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -886,7 +886,9 @@ function dot(x::AbstractArray, y::AbstractArray)
     s
 end
 
-dot(x::Adjoint, y::Adjoint) = conj(dot(parent(x), parent(y)))
+function dot(x::Adjoint{<:Union{Real,Complex}}, y::Adjoint{<:Union{Real,Complex}})
+    return conj(dot(parent(x), parent(y)))
+end
```

Comment:
You could add another method for the generic `Number` case to extend the performance optimization to `Quaternion`s.
```suggestion
function dot(x::Adjoint{<:Union{Real,Complex}}, y::Adjoint{<:Union{Real,Complex}})
    return conj(dot(parent(x), parent(y)))
end
function dot(x::Adjoint{<:Number}, y::Adjoint{<:Number})
    return conj(dot(parent(y), parent(x)))
end
```

Reply:
Doesn't this definition work as the general fallback?

Reply:
🤦  Of course. I noticed that we need to restrict to `Numbers`, because for arrays of arrays the `conj`s seem to annihilate each other and the result becomes wrong.

Reply:
This overload is also not correct:
```julia
julia> using LinearAlgebra, Quaternions

julia> function LinearAlgebra.dot(x::Adjoint{<:Number}, y::Adjoint{<:Number})
           return conj(dot(parent(y), parent(x)))
       end

julia> x = [quatrand() for _ in 1:2, _ in 1:2];

julia> y = [quatrand() for _ in 1:2, _ in 1:2];

julia> dot(Matrix(x'), Matrix(y'))
Quaternion{Float64}(11.624166294288848, 5.2859863153466815, -2.9677753842907677, -4.862662050905291, false)

julia> dot(x', y')
Quaternion{Float64}(11.624166294288848, 4.145847727927628, -2.535886196576731, -3.489570292291287, false)
```

This is because we want to compute
`⟨xᴴ, yᴴ⟩ = ∑ᵢ (xᵀ)ᵢ (yᵀ)ᵢ^* = ∑ᵢ xᵢ yᵢ^*`
But the overload is actually computing
`⟨y, x⟩^* = (∑ᵢ yᵢ^* xᵢ)^* = ∑ᵢ xᵢ^* yᵢ`

We could compute what we want to by conjugating both input arguments:
`⟨x^*, y^*⟩ = ∑ᵢ xᵢ yᵢ^*`
But then we end up allocating 2 arrays.